File: Java/heap/MaxItemWithGivenSum.java
package heap;

import java.util.Arrays;

public class MaxItemWithGivenSum {
    /**
     * Time Complexity -> O(n * logn)
     * <p>
     * Auxiliary Space -> O(1)
     * 
     * @param arr Array of items cost
     * @param sum Total amount in the hand
     * @return Number of item that can be purchased
     */
    public static int maxItemUsingSortedArray(int[] arr, int sum) {
        int count = 0;
        Arrays.sort(arr);
        for (int i : arr) {
            if (i > sum)
                return count;
            count++;
            sum -= i;
        }
        return count;
    }

    /**
     * Time Complexity -> O(n * logn)
     * <p>
     * Auxiliary Space -> O(n)
     * 
     * @param arr Array of items cost
     * @param sum Total amount in the hand
     * @return Number of item that can be purchased
     */
    public static int maxItemUsingHeap(int[] arr, int sum) {
        PriorityQueue pq = new PriorityQueue();
        for (int i : arr)
            pq.add(i);
        int count = 0;
        while (!pq.isEmpty() && pq.peek() <= sum) {
            sum -= pq.poll();
            count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int[] arr = { 1, 12, 5, 111, 200 };
        System.out.println(maxItemUsingSortedArray(arr, 11));
        System.out.println(maxItemUsingHeap(arr, 11));
    }
}

File: Python/heap/max_item_with_given_sum.py
import priority_queue as pq


def max_item_using_sorted_array(arr, sum_):
    count = 0
    for i in sorted(arr):
        if i > sum_:
            return count
        count += 1
        sum_ -= i
    return count


def max_item_using_heap(arr, sum_):
    for i in arr:
        pq.add(i)
    count = 0
    while not pq.is_empty() and pq.peek() <= sum_:
        sum_ -= pq.poll()
        count += 1
    return count


def main():
    arr = [1, 12, 5, 111, 200]
    print(max_item_using_sorted_array(arr, 11))
    print(max_item_using_heap(arr, 11))


if __name__ == "__main__":
    main()
